### ISSUE TITLE
ENH/DOC: Timestamp/DTI to epoch time

### ISSUE THREAD

**Issue by jreback** [pandas-dev/pandas#14772]
Converting a Timestamp/DTI to epoch time involves a little recipe. We could make this a bit simpler in a couple of ways:

a) document  the below
b) provide a ``.to_epoch(unit=)`` method on Timestamp/DTI/.dt
c) allow an enhanced astype, IOW, ``.astype('int64[s]')
d) allow ``.to_numeric(...., unit=)``
e) allow Timestamp/DTI division with a Timedelta, e.g. ``s / Timedelta('1s')`` 

```
In [19]: s = Series(pd.date_range('20160101',periods=3))

In [20]: s
Out[20]: 
0   2016-01-01
1   2016-01-02
2   2016-01-03
dtype: datetime64[ns]

In [21]: ((s-Timestamp(0)) / Timedelta('1s')).astype('i8')
Out[21]: 
0    1451606400
1    1451692800
2    1451779200
dtype: int64

# this is the inverse
In [25]: pd.to_datetime(((s-Timestamp(0)) / Timedelta('1s')).astype('i8'), unit='s')
Out[25]: 
0   2016-01-01
1   2016-01-02
2   2016-01-03
dtype: datetime64[ns]
```

cc @sinhrks @jorisvandenbossche @chris-b1

**Comment by jreback** [comment]
xref #11022, https://github.com/pandas-dev/pandas/issues/6741

**Comment by jreback** [comment]
This also works, but exposing internal impl, verbose and not user friendly
```
In [36]: Series(s.values.astype('datetime64[s]').astype('i8'), index=s.index)
Out[36]: 
0    1451606400
1    1451692800
2    1451779200
dtype: int64
```

**Comment by jorisvandenbossche** [comment]
If we would add user facing functionality, I think I would like `to_epoch()` most (certainly not something like `int64[s]` IMO)

**Comment by jreback** [comment]
I changed this to make this an enhancement for a simple ``.to_epoch()`` method on Timestamp/DTI.

**Comment by jbrockmendel** [comment]
since timestamp now has timestamp method, should,we use the same name for DTI?

**Comment by jreback** [comment]
yes this would be reasonable (though to be honest the ``.timestamp()`` name is not very informative.....

**Comment by jorisvandenbossche** [comment]
I also don't really like the name. It is rather confusing given that we already have a `Timestamp` class (for timestamps itself it is ok to keep subclass consistency). So when adding such a method to DatetimeIndex / dt accessor, I would think about not using the same name.

**Comment by jreback** [comment]
I am partial to ``to_epoch``, we use this term elsewhere.

**Comment by TomAugspurger** [comment]
One question: what to do with NaTs?

```python
In [5]: pd.DatetimeIndex(['2017', '2018', None]).values.astype('datetime64[s]').astype("i8")
Out[5]: array([          1483228800,           1514764800, -9223372036854775808])
```

Do we value having integer dtype more? I think so in this case.

**Comment by TomAugspurger** [comment]
Second question: timezones. Unix time is defined in UTC, so should we

- assume that TZ-naive data is UTC, and just localize
- convert TZ-aware data to UTC

This will nescessitate an `ambiguous` parameter.

**Comment by TomAugspurger** [comment]
third question: how to handle higher-precision components?

```python
In [6]: pd.DatetimeIndex(['2017-01-01T00:00:00.01', '2017-01-01T00:00:00.02']).to_epoch()
Out[6]: array([1483228800, 1483228800])
```

I don't think we should use floats and fractional components. So that leaves truncating or rounding to the nearest `unit`.

**Comment by simonjayhawkins** [comment]
> One question: what to do with NaTs?
> 
> ```python
> In [5]: pd.DatetimeIndex(['2017', '2018', None]).values.astype('datetime64[s]').astype("i8")
> Out[5]: array([          1483228800,           1514764800, -9223372036854775808])
> ```
> 
> Do we value having integer dtype more? I think so in this case.

could a `<IntegerArray>` be returned in this case. it would need some casting since currently

```python
pd.array(pd.DatetimeIndex(['2017', '2018', None]).values.astype('datetime64[s]'), dtype='Int64')
```
raises
```
TypeError: datetime64[s] cannot be converted to an IntegerDtype
```

**Comment by jreback** [comment]
yes there are various use cases where we could do things like this
